pkg/clusteragent/admission/patch/patcher.go

```diff
@@ -0,0 +1,104 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+// +build kubeapiserver
+
+package patch
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/admission/common"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+)
+
+type patcher struct {
+	k8sClient        kubernetes.Interface
+	isLeader         func() bool
+	deploymentsQueue chan PatchRequest
+}
+
+func newPatcher(k8sClient kubernetes.Interface, isLeaderFunc func() bool, pp patchProvider) *patcher {
+	return &patcher{
+		k8sClient:        k8sClient,
+		isLeader:         isLeaderFunc,
+		deploymentsQueue: pp.subscribe(KindDeployment),
+	}
+}
+
+func (p *patcher) start(stopCh <-chan struct{}) {
+	for {
+		select {
+		case req := <-p.deploymentsQueue:
+			if err := p.patchDeployment(req); err != nil {
+				log.Error(err.Error())
+			}
+		case <-stopCh:
+			log.Info("Shutting down patcher")
+			return
+		}
+	}
+}
+
+func (p *patcher) patchDeployment(req PatchRequest) error {
+	if !p.isLeader() {
+		log.Debug("Not leader, skipping")
+		return nil
+	}
+	deploy, err := p.k8sClient.AppsV1().Deployments(req.K8sTarget.Namespace).Get(context.TODO(), req.K8sTarget.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	oldObj, err := json.Marshal(deploy)
+	if err != nil {
+		return fmt.Errorf("failed to encode object: %v", err)
+	}
+	versionAnnotKey := fmt.Sprintf(common.LibVersionAnnotKeyFormat, req.LibID.Language)
+	configAnnotKey := fmt.Sprintf(common.LibConfigAnnotKeyFormat, req.LibID.Language)
+	switch req.Action {
+	case ApplyConfig:
+		if deploy.Spec.Template.Labels == nil {
+			deploy.Spec.Template.Labels = make(map[string]string)
+		}
+		deploy.Spec.Template.Labels[common.EnabledLabelKey] = "true"
+		if deploy.Spec.Template.Annotations == nil {
+			deploy.Spec.Template.Annotations = make(map[string]string)
+		}
+		deploy.Spec.Template.Annotations[versionAnnotKey] = req.LibID.Version
+		conf, err := json.Marshal(req.LibConfig)
+		if err != nil {
+			return fmt.Errorf("failed to encode library config: %v", err)
+		}
+		deploy.Spec.Template.Annotations[configAnnotKey] = string(conf)
+	case DisableInjection:
+		if deploy.Spec.Template.Labels == nil {
+			deploy.Spec.Template.Labels = make(map[string]string)
+		}
+		deploy.Spec.Template.Labels[common.EnabledLabelKey] = "false"
```

Comment:
In addition to better debuggability, if we just delete the label we don't guarantee disabling it because of the mutate enabled config. See this
![image](https://user-images.githubusercontent.com/38987709/205599885-5c3298ce-995e-4782-b6e9-c95910a332fb.png)

Reply:
oh wow, is this something we should have a comment about or would someone maintaining this be familiar enough to know?